_posts/2023-10-11-kafka-consumer-client-essentials.md

```diff
@@ -0,0 +1,485 @@
+* `AUTO_OFFSET_RESET_CONFIG` to define the starting point for consuming messages when no committed offset is found for the configured group ID and assigned partition.  
+* `KEY_DESERIALIZER_CLASS_CONFIG` and `VALUE_DESERIALIZER_CLASS_CONFIG` to specify deserializers that transform message keys and values into a suitable format. 
+In this case, we'll specify the `ByteArrayDeserializer` because it simply represents the keys and values as bytes without further transformation.
+
+> Using the earliest offset behavior avoids messages being lost when the offsets retention period (`offsets.retention.minutes`) configured for a broker has ended.
+
+Producers and consumers can use different serializers and deserializers as long as the consumer's deserializer is capable of correctly interpreting the serialized data produced by the producer's serializer.
+In our [producer example](https://strimzi.io/blog/2023/10/03/kafka-producer-client-essentials/), messages are generated as random byte arrays, which serve as the content of the messages sent to the Kafka cluster. Given this format, it makes sense to choose the `ByteArrayDeserializer` for deserialization.  
+
+We'll also include methods that help with these operations:
+
+**`sleep` method**
+
+- Introduces a delay in the message-receiving process for a specified number of milliseconds.
+- Useful for controlling the message consumption rate or simulating message processing times.
+- Handles potential `InterruptedException` if the thread responsible for sending messages is interrupted while paused.
+
+**`retriable` method**
+
+- Determines whether to retry message consumption and offset management following a message consuming error.
+- Returns `true` if the message consuming process can be retried. The Kafka consumer automatically handles retries for certain errors, such as connection errors.
+- Returns `false` for null and specified exceptions, or those that do not implement the `RetriableException` interface.
+- Customizable to include other errors and implementing retry logic for business level exceptions.
+
+> By default, Kafka operates with at-least-once message delivery semantics, which means that messages can be delivered more than once in certain scenarios, potentially leading to duplicates.
+> Consider using transactional ids and enabling idempotence (`enable.idempotence=true`) on the producer side to guarantee exactly-once delivery. 
+> On the consumer side, you can then use the `isolation.level` property to control how transactional messages are read by the consumer.
+> For more information, see the [Strimzi post on transactions](https://strimzi.io/blog/2023/05/03/kafka-transactions/)
+
+**`onPartitionsAssigned` method**
+
+- Provides information about the partitions assigned to the consumer.
+- Messages are not consumed during partition assignment. However, constantly slow or failed assignments could indicate cluster or network issues. 
```

Comment:
Not sure `However` works in this sentence. Also not sure what you mean by `slow or failed assignments`?

Reply:
I think this change came from an earlier review. I've adjusted as follows:

- Messages are not consumed during partition assignment. If there is a delay or failure in the consumer being assigned to partitions, this can disrupt the consumption process.